packages/compass-schema-validation/src/modules/sample-documents.ts

```diff
@@ -0,0 +1,307 @@
+});
+
+const ACTION_TO_REDUCER_MAPPINGS: {
+  [Type in SampleDocumentAction['type']]: (
+    state: SampleDocumentState,
+    action: SampleDocumentAction & { type: Type }
+  ) => SampleDocumentState;
+} = {
+  [CLEAR_SAMPLE_DOCUMENTS]: clearingSampleDocuments,
+  [FETCHING_VALID_DOCUMENT]: startFetchingValidDocument,
+  [FETCHED_VALID_DOCUMENT]: updateStateWithFetchedValidDocument,
+  [FETCHING_VALID_DOCUMENT_FAILED]: validDocumentFetchErrored,
+  [FETCHING_INVALID_DOCUMENT]: startFetchingInvalidDocument,
+  [FETCHED_INVALID_DOCUMENT]: updateStateWithFetchedInvalidDocument,
+  [FETCHING_INVALID_DOCUMENT_FAILED]: invalidDocumentFetchErrored,
+};
+
+export default function (
+  state: SampleDocumentState = INITIAL_STATE,
+  action: RootAction
+): SampleDocumentState {
+  const fn =
+    ACTION_TO_REDUCER_MAPPINGS[action.type as SampleDocumentAction['type']];
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore-error TS does not understand that action can be passed to fn
+  return fn ? fn(state, action) : state;
+}
+
+/**
+ * Side effects
+ */
+
+const getSampleDocuments = async ({
+  pipeline,
+  namespace,
+  dataService,
+}: {
+  namespace: string;
+  dataService: DataService;
+  pipeline: Record<string, unknown>[];
+}) => {
+  const aggOptions = {
+    allowDiskUse: true,
+    maxTimeMS: preferences.getPreferences().maxTimeMS,
+  };
+  pipeline.unshift({ $sample: { size: SAMPLE_SIZE } });
+
+  return await dataService.aggregate(namespace, pipeline, aggOptions);
+};
+
+export const fetchValidDocument = () => {
+  return async (
+    dispatch: (action: RootAction) => void,
+    getState: () => RootState
+  ) => {
+    dispatch(fetchingValidDocument());
+
+    const state = getState();
+    const dataService = state.dataService;
+    const namespace = state.namespace.ns;
+    const validator = state.validation.validator;
+
+    const checkedValidator = checkValidator(validator);
+    const query = checkValidator(
+      checkedValidator.validator as string
+    ).validator;
+
+    if (!dataService) {
+      return;
+    }
+
+    try {
+      const valid = (
+        await getSampleDocuments({
+          namespace,
+          dataService,
+          pipeline: [{ $match: query }, { $limit: 1 }],
+        })
+      )[0];
+
+      dispatch(fetchedValidDocument(valid));
+    } catch (e) {
+      dispatch(fetchingValidDocumentFailed());
+    }
+  };
+};
+
+export const fetchInvalidDocument = () => {
+  return async (
+    dispatch: (action: RootAction) => void,
+    getState: () => RootState
+  ) => {
+    dispatch(fetchingInvalidDocument());
+
+    const state = getState();
+    const dataService = state.dataService;
+    const namespace = state.namespace.ns;
+    const validator = state.validation.validator;
+
+    // Calling checkValidator twice here seems wrong
+    const checkedValidator = checkValidator(validator);
+    const query = checkValidator(
+      checkedValidator.validator as string
+    ).validator;
```

Comment:
Added the comment here, something can’t be quite right here but the code was admittedly already a bit hard to follow because it somewhat freely mixes validators in string representation vs JS object representation